src/Storages/StorageMergeTree.cpp

```diff
@@ -688,8 +694,25 @@ bool StorageMergeTree::merge(
         new_part = merger_mutator.mergePartsToTemporaryPart(
             future_part, metadata_snapshot, *merge_entry, table_lock_holder, time(nullptr),
             merging_tagger->reserved_space, deduplicate, force_ttl);
+
         merger_mutator.renameMergedTemporaryPart(new_part, future_part.parts, nullptr);
 
+        DataPartsVector parts_to_remove_immediately;
```

Comment:
I think we need to move this logic into `grabOldParts` of `MergeTree`. Maybe introduce additional setting for lifetime of in-memory parts.